src/theory/splitter.h

```diff
@@ -0,0 +1,136 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Amalee Wilson, Andrew Wu
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Splitter for creating partitions.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__SPLITTER_H
+#define CVC5__THEORY__SPLITTER_H
+
+#include <fstream>
+#include <vector>
+#include <sstream>
+
+#include "proof/trust_node.h"
+#include "theory/valuation.h"
+
+namespace cvc5 {
+
+class TheoryEngine;
+
+namespace prop {
+class PropEngine;
+}
+
+namespace theory {
+
+class Splitter
+{
+ public:
+  Splitter(TheoryEngine* theoryEngine, prop::PropEngine* propEngine);
+
+  /**
+   * Make partitions for parallel solving.
+   */
+  TrustNode makePartitions();
+
+ private:
+  /**
+   * Print the cube to the specified output.
+   */
+  void emitCube(Node toEmit);
+
+  /**
+   * Block a path in the search by sending the not of toBlock as a lemma to the
+   * SAT solver.
+   */
+  TrustNode blockPath(Node toBlock);
+
+  /**
+   * Close the output file if it is specified.
+   */
+  void closeFile();
+
+  /**
+   * Stop partitioning and return unsat.
+   */
+  TrustNode stopPartitioning();
+
+  /**
+   * Get the list of decisions from the SAT solver
+   */
+  void collectDecisionLiterals(std::vector<TNode>& literals);
+
+  /**
+   * Current propEngine.
+   */
+  prop::PropEngine* d_propEngine;
+
+  /**
+   * Valuation of the theory engine.
+   */
+  std::unique_ptr<Valuation> d_valuation;
+
+  /**
+   * The number of partitions requested through the compute-partitions option.
+   */
+  const uint16_t d_numPartitions;
```

Comment:
Is there a specific reason to use only 16 bits here? Note that the option itself is `uint64_t`, and there is no reason to use anything less than 64 bit in most cases... (same for the other members)

Reply:
Couldn't I also change the option itself? Seems unlikely that you'd request enough partitions to need 64 bits. But I don't mind changing it.